### ISSUE TITLE
NgModelChange will be called after validate

### ISSUE THREAD

**Issue by lexon0011** [angular/angular#24067]
## I'm submitting a...
<pre><code>
[X] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
Since Angular 6 ngModelChange is emitted after the value/validity is updated on its control, which breaks our custom validation. 

## Expected behavior
Our custom validation method uses a bound model to process the validation. Because the validate method is called before the current change is propagated the custom validation result is unexpected, because we rely on current values on our model.

## Minimal reproduction of the problem with instructions
Sample: [https://stackblitz.com/edit/angular-cmj6jf](https://stackblitz.com/edit/angular-cmj6jf)

As you can see in the sample, the custom validation method `validateTest()` uses the value of the bound property, which is not the current value of the form control.

I create a second sample with angular 5.5.10, which works as expected: [https://stackblitz.com/edit/angular-pcxosg](https://stackblitz.com/edit/angular-pcxosg)

## Questions:
- Is there a solution for our problem?
- Maybe is it possible to trigger programmatically a modelChange in the `CustomValidationDirective`? 

## Angular version: 
6.0.3

**Comment by kara** [comment]
This is actually an intended breaking change. See the explanation / PR [here](https://github.com/angular/angular/pull/21514#issuecomment-367524768).

Rather than relying on a property on your class being set, I'd use the [control value](https://stackblitz.com/edit/angular-cmj6jf?file=src%2Fapp%2Fapp.component.ts) that is passed into the validation function. This way, the value will always be up-to-date and it doesn't rely on the internal order of execution.

**Comment by rehfeldchris** [comment]
Ya this breaking change was a bit disappointing. Not the fact that it was fixed, but that it wasn't communicated to upgraders very well. 

I saw the ngModelChange notice in the upgrade guide, but I wasn't aware that it would affect me, as I don't explicitly use ngModelChange anywhere. But, I do use custom validators, and I use databinding, eg
` <input  [(ngModel)]="userIdVal"... `

and then in my validators I use the `userIdVal `variable. But, this is broken, and the `userIdVal `is always 1 character behind the latest keystroke because of this change because the validators get run before the `userIdVal `gets updated with the latest value. 

I wish I would have known how much more breadth this breaking change really caused. I think users would appreciate if an expert were to document it better in the upgrade guide.
